### PR TITLE
TPC-H benchmark

### PR DESCRIPTION
**Pull request by majetideepak**
The scope of this PR is to build a benchmark for TPC-H plans in Velox.
Both DWRF and Parquet file formats are supported. Since DWRF does not support the `Date` type, the filters are adjusted accordingly.
The TPC-H data must be specified using the `data_path` option.
Each table data must be placed in a hive-style partition.
Sample invocation
```
cmake-build-release/velox/benchmarks/velox_tpch_benchmark --num_drivers 16 --data_format parquet --data_path /data/tpch10
============================================================================
velox/benchmarks/TpchBenchmark.cpprelative  time/iter  iters/s
============================================================================
q1                                          2.93s  340.75ms
q6                                         926.40ms     1.08
q18                                        12.56s   79.65m
============================================================================

cmake-build-release/velox/benchmarks/velox_tpch_benchmark --num_drivers 16 --data_format orc --data_path /data/tpch10
============================================================================
velox/benchmarks/TpchBenchmark.cpprelative  time/iter  iters/s
============================================================================
q1                                          1.81s      553.58m
q6                                          566.12ms     1.77
q18                                        9.21s  108.53m
============================================================================
```
The benchmark also supports `run_query_verbose` option that can be used to print the plan along with execution statistics for a particular query. 
```
// Default num_drivers = 4
cmake-build-release/velox/benchmarks/velox_tpch_benchmark --run_query_verbose 6 --data_path /data/tpch10
Execution time: 1445 ms
Splits total: 80, finished: 80
-> Aggregation[FINAL a0 := sum(ROW["a0"])]
   Output: 0 rows (0 bytes), Cpu time: 0ns, Blocked wall time: 0ns, Peak memory: 0 bytes
  -> LocalPartition[]
     Output: 1 rows (32 bytes), Cpu time: 68663ns, Blocked wall time: 66402000ns, Peak memory: 0 bytes
     LocalPartition: Output: 1 rows (32 bytes), Cpu time: 68663ns, Blocked wall time: 66402000ns, Peak memory: 0 bytes
        queuedWallNanos    sum: 147000, count: 2, min: 53000, max: 94000
     LocalExchangeSource: Output: 0 rows (0 bytes), Cpu time: 0ns, Blocked wall time: 0ns, Peak memory: 0 bytes
    -> Aggregation[PARTIAL a0 := sum(ROW["p0"])]
       Output: 2 rows (64 bytes), Cpu time: 6040109ns, Blocked wall time: 0ns, Peak memory: 1048576 bytes
      -> Project[expressions: (p0:DOUBLE, multiply(ROW["l_extendedprice"],ROW["l_discount"])), ]
         Output: 845534 rows (9096355 bytes), Cpu time: 1251764323ns, Blocked wall time: 0ns, Peak memory: 1048576 bytes
            dataSourceLazyWallNanos    sum: 1245148000, count: 4462, min: 1000, max: 4616000
        -> Project[expressions: (l_shipdate:VARCHAR, ROW["shipdate"]), (l_extendedprice:DOUBLE, ROW["extendedprice"]), (l_quantity:DOUBLE, ROW["quantity"]), (l_discount:DOUBLE, ROW["discount"]), ]
           Output: 845534 rows (683555500 bytes), Cpu time: 13060400ns, Blocked wall time: 0ns, Peak memory: 0 bytes
          -> TableScan[]
             Input: 845534 rows (1047829184 bytes), Raw Input: 44568184 rows (425846631 bytes), Output: 845534 rows (683555500 bytes), Cpu time: 2724025605ns, Blocked wall time: 220000ns, Peak memory: 20971520 bytes
                dataSourceWallNanos    sum: 2786619000, count: 4520, min: 5000, max: 29078000
                ramReadBytes           sum: 0, count: 3, min: 0, max: 0
                queuedWallNanos        sum: 199000, count: 3, min: 34000, max: 105000
                localReadBytes         sum: 0, count: 3, min: 0, max: 0
                numLocalRead           sum: 0, count: 3, min: 0, max: 0
                numStorageRead         sum: 0, count: 3, min: 0, max: 0
                skippedSplitBytes      sum: 0, count: 3, min: 0, max: 0
                skippedStrides         sum: 0, count: 3, min: 0, max: 0
                numPrefetch            sum: 0, count: 3, min: 0, max: 0
                numRamRead             sum: 0, count: 3, min: 0, max: 0
                prefetchBytes          sum: 0, count: 3, min: 0, max: 0
                skippedSplits          sum: 0, count: 3, min: 0, max: 0
                storageReadBytes       sum: 0, count: 3, min: 0, max: 0
```

### CODE REVIEW

**Comment by majetideepak** [comment]
@9prady9, @mbasmanova, @pedroerp Just a heads-up that after https://github.com/facebookincubator/velox/pull/1073 gets merged, I plan to refactor the TPC-H plans into their own class and re-use them inside both ParquetTpchTest and this TPC-H Benchmark class that I am implementing.
We can benchmark TPC-H queries over a bigger scale factor periodically.

**Comment by majetideepak** [comment]
@pedroerp This benchmark is currently disabled in CI. We can enable it if we add TPC-H data generation as part of the CI job.

**Comment by majetideepak** [comment]
@9prady9 I compared q1 against DuckDB on my local laptop for 16 threads and parquet data for SF10. The time reported by DuckDB was Total Time: 2.48s. So Velox Parquet was slower, but DWRF was faster.
Let me know your thoughts on this refactor as well.

File: velox/type/Variant.h [hunk 1]
```diff
@@ -29,7 +29,7 @@
 namespace facebook::velox {
 
 // Constant using in comparison of REAL and DOUBLE values.
-constexpr double kEpsilon{0.000001};
+constexpr double kEpsilon{0.00001};
```
Comment (mbasmanova) [inline velox/type/Variant.h:32]: @majetideepak Perhaps, move this change into a separate PR. What's the motivation for it?

File: velox/exec/tests/utils/HiveConnectorTestBase.h [hunk 2]
```diff
@@ -84,6 +84,16 @@ class HiveConnectorTestBase : public OperatorTestBase {
     return makeHiveConnectorSplit(filePath, {}, start, length);
   }
 
+  // Split file at a given path 'filePath' in range [start, length] into
+  // 'numSplits' splits.
+  static std::vector<std::shared_ptr<connector::hive::HiveConnectorSplit>>
+  makeHiveSplitsByCount(
```
Comment (mbasmanova) [inline velox/exec/tests/utils/HiveConnectorTestBase.h:90]: A few observations and suggestions:
-  Other similar methods are named makeConnectorSplit. It would be nice to rename this method to be consistent. 
- This method relies on length to be the actual size of the file, while other methods accept arbitrary large lengths. Given that the usage of this method is to always pass zero for start and file size for length, it might be helpful to change the method signature to remove start and length parameters. 
- The default value for splitCount parameter seems arbitrary. Consider removing it and requiring the caller to specify number of splits.

```
static std::vector<std::shared_ptr<connector::hive::HiveConnectorSplit>>
makeHiveConnectorSplits(
   const std::string& filePath, 
   uint32_t splitCount, 
   dwio::common::FileFormat format);
```

File: velox/benchmarks/SimpleArithmetic.cpp [hunk 3]
```diff
@@ -15,13 +15,11 @@
  */
 
 #include <folly/Benchmark.h>
-#include <folly/init/Init.h>
```
Comment (mbasmanova) [inline velox/benchmarks/SimpleArithmetic.cpp:18]: These changes seem unrelated to this diff. It would be nice to move them into a separate PR.

File: velox/type/Variant.h (hunk 1 above)
Comment (majetideepak) [reply velox/type/Variant.h:32]: @mbasmanova, I removed the explicit `round()` from the plans since we now have support for REAL and float results comparison. One value still failed between DuckDB and Velox due to precision difference. This fixed the problem. Makes sense to move it to another PR.

File: velox/exec/tests/utils/TpchBuilder.h [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+/// Since the column names in the file can vary, they are mapped to the standard
+/// names. Therefore, the order of the columns in the file is important and
+/// should be in the same order as in the TPC-H standard.
+class TpchBuilder {
+ public:
+  /// Read each data file and initialize row types and determine data paths for
+  /// each table.
+  /// @param dataPath path to the data files
+  /// @param format file format
+  void initialize(std::string dataPath, std::string format) {
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchBuilder.h:70]: @majetideepak Would you move the implementation to a .cpp file for the sake of faster builds?

File: Makefile [hunk 5]
```diff
@@ -88,12 +88,12 @@ min_debug:				#: Minimal build with debugging symbols
 	$(MAKE) build BUILD_DIR=debug
 
 benchmarks-build:
-	$(MAKE) release EXTRA_CMAKE_FLAGS="-DVELOX_BUILD_MINIMAL=ON -DVELOX_BUILD_BENCHMARKS=ON"
+	$(MAKE) release EXTRA_CMAKE_FLAGS=" -DVELOX_ENABLE_PARQUET=ON -DVELOX_BUILD_BENCHMARKS=ON"
```
Comment (mbasmanova) [inline Makefile:91]: I wonder if it makes sense to keep a lean benchmark and add a new one for TPC-H. E.g. we could have a quick to build and run benchmark for core building blocks available in Velox and a separate longer to build and run benchmark that runs complex queries like TPC-H and TPC-DS.

File: velox/exec/tests/utils/TpchBuilder.h [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+      case 1:
+        return getTpchQ1Plan();
+      case 6:
+        return getTpchQ6Plan();
+      default:
+        VELOX_NYI();
+    }
+  }
+
+  PlanContext getTpchQ1Plan() const {
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchBuilder.h:123]: This method can be private. The callers would use getTpchPlan(1).

File: velox/exec/tests/utils/TpchBuilder.h [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+                    .finalAggregation({}, {"sum(a0)"}, {DOUBLE()})
+                    .planNode();
+    PlanContext context;
+    context.plan_ = std::move(plan);
+    context.sources_[lineitemPlanNodeId] = getTableFilePaths(kTableName);
+    context.format_ = format_;
+    return context;
+  }
+
+  const std::vector<std::string>& getTableFilePaths(
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TpchBuilder.h:284]: This method can be private as well.

File: velox/exec/tests/utils/TpchBuilder.h (hunk 4 above)
Comment (majetideepak) [reply velox/exec/tests/utils/TpchBuilder.h:70]: I wanted to avoid an additional library(`velox-exec-test-util`) dependency by keeping it header only. I don't have a strong opinion here and can move it to a `.cpp` file if that is preferred.

File: Makefile (hunk 5 above)
Comment (majetideepak) [reply Makefile:91]: I agree.  Will separate this out.

File: velox/exec/tests/utils/TpchBuilder.h (hunk 4 above)
Comment (mbasmanova) [reply velox/exec/tests/utils/TpchBuilder.h:70]: Let's move to .cpp

**Comment by majetideepak** [comment]
@mbasmanova can you take another look? thanks.

**Comment by majetideepak** [comment]
@mbasmanova thanks very much! I will move the TpchQueryBuilder to a separate one.

File: velox/dwio/parquet/tests/ParquetTpchTest.cpp [hunk 8]
```diff
@@ -58,209 +57,85 @@ class ParquetTpchTest : public testing::Test {
...
-        tableName,
-        filePath,
-        kRowGroupSize);
-    duckDb_->execute(query);
-    return filePath;
+    const auto tableNames = tpchBuilder_.getTableNames();
+    for (const auto& tableName : tableNames) {
+      auto tableDirectory =
+          fmt::format("{}/{}", tempDirectory_->path, tableName);
+      fs::create_directory(tableDirectory);
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/dwio/parquet/tests/ParquetTpchTest.cpp:72]: Since we are writing inside a temp directory, the `TempDirectoryPath` destructor ensures the cleanup.

File: velox/exec/tests/utils/TpchBuilder.h [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  /// Table type inferred from the input data file.
+  std::map<std::string, RowTypePtr> tableTypes_;
+  /// File paths for each Table.
+  std::unordered_map<std::string, std::vector<std::string>> dataPaths_;
+  /// Mapping standard column name to table type column name.
+  std::unordered_map<std::string, std::unordered_map<std::string, std::string>>
+      columnAlias_;
+  dwio::common::FileFormat format_;
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/exec/tests/utils/TpchBuilder.h:113]: It could if we add a constructor for initialization. Will add one.

File: velox/exec/PlanNodeStats.cpp [hunk 10]
```diff
@@ -55,8 +55,8 @@ std::string PlanNodeStats::toString(bool includeInputStats) const {
         << "), ";
   }
   out << "Output: " << outputRows << " rows (" << outputBytes << " bytes)"
-      << ", Cpu time: " << cpuWallTiming.cpuNanos << "ns"
-      << ", Blocked wall time: " << blockedWallNanos << "ns"
+      << ", Cpu time: " << cpuWallTiming.cpuNanos * 1.0 / 1'000 << "ms"
```
Comment (mbasmanova) [inline velox/exec/PlanNodeStats.cpp:58]: We need to add proper logic to format these numbers by checking how large the number is and decidig whether to report it using ns, ms, s or min.

File: velox/CMakeLists.txt [hunk 11]
```diff
@@ -64,3 +64,7 @@ endif()
 if(${VELOX_CODEGEN_SUPPORT})
   add_subdirectory(experimental/codegen)
 endif()
+
+if(${VELOX_ENABLE_BENCHMARKS_LARGE})
+  add_subdirectory(benchmarks_large)
```
Comment (mbasmanova) [inline velox/CMakeLists.txt:69]: I'm expecting more benchmarks and wonder if we could put all of them under benchmarks folder. TPC-H could go under benchmarks/tpch.

File: velox/benchmarks_large/TpchBenchmark.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+static bool notEmpty(const char* /*flagName*/, const std::string& value) {
+  return !value.empty();
+}
+} // namespace
+
+DEFINE_string(data_path, "", "Root path of TPC-H data");
+DEFINE_int32(query, 1, "Query number");
+DEFINE_int32(num_drivers, 4, "Number of drivers");
+DEFINE_string(format, "parquet", "Data format");
```
Comment (mbasmanova) [inline velox/benchmarks_large/TpchBenchmark.cpp:46]: It would be helpful to specify what values are supported.

File: velox/benchmarks_large/TpchBenchmark.cpp [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+DEFINE_bool(verbose, false, "Print benchmark results in detail");
+
+DEFINE_validator(data_path, &notEmpty);
```
Comment (mbasmanova) [inline velox/benchmarks_large/TpchBenchmark.cpp:49]: Should we add a validator for 'format'?

File: velox/benchmarks_large/TpchBenchmark.cpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+            ->newConnector(kHiveConnectorId, nullptr);
+    connector::registerConnector(hiveConnector);
+  }
+
+  std::shared_ptr<Task> run(const TpchPlan& tpchPlan) {
+    CursorParameters params;
+    params.maxDrivers = FLAGS_num_drivers;
+    params.numResultDrivers = 1;
+    params.planNode = tpchPlan.plan;
+    constexpr int kNumSplits = 10;
```
Comment (mbasmanova) [inline velox/benchmarks_large/TpchBenchmark.cpp:70]: This is number of splits per file, not total number of splits, right? If so, consider renaming to kNumSplitsPerFile for clarity. Would it make sense to make this configurable?

File: velox/benchmarks_large/TpchBenchmark.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+  const auto planContext = queryBuilder.getQueryPlan(6);
+  benchmark.run(planContext);
+}
+
+int main(int argc, char** argv) {
+  folly::init(&argc, &argv, false);
+  benchmark.initialize();
+  queryBuilder.initialize(FLAGS_data_path);
+  if (!FLAGS_verbose) {
+    folly::runBenchmarks();
```
Comment (mbasmanova) [inline velox/benchmarks_large/TpchBenchmark.cpp:111]: It looks like in this case FLAGS_query parameter is ignored. Is this expected?

**Review by mbasmanova — COMMENTED** [review]
@majetideepak Looks nice. Some comments below. I'm seeing a lot of "Peak memory: 0 bytes" in the results. https://github.com/facebookincubator/velox/pull/1174 should help fix that.

File: velox/exec/PlanNodeStats.cpp (hunk 10 above)
Comment (majetideepak) [reply velox/exec/PlanNodeStats.cpp:58]: Makes sense. I will look into this logic in a separate PR. Will revert this for now.

File: velox/benchmarks_large/TpchBenchmark.cpp (hunk 15 above)
Comment (majetideepak) [reply velox/benchmarks_large/TpchBenchmark.cpp:111]: Yes, `FLAGS_query` is only used for `verbose` mode. All the tpch queries are executed in the benchmark mode.

File: velox/benchmarks_large/TpchBenchmark.cpp (hunk 14 above)
Comment (majetideepak) [reply velox/benchmarks_large/TpchBenchmark.cpp:70]: will fix.

File: velox/CMakeLists.txt (hunk 11 above)
Comment (mbasmanova) [reply velox/CMakeLists.txt:69]: Would it make sense to move current code in benchmarks into benchmarks/basic ? I feel that having too many top-level directories makes it harder to browse the code base.

File: velox/benchmarks_large/TpchBenchmark.cpp (hunk 15 above)
Comment (mbasmanova) [reply velox/benchmarks_large/TpchBenchmark.cpp:111]: Wouldn't this be surprising to users of the benchmark? Usually, verbose flag only affected the output of the program, not what the program does.

Comment (majetideepak) [reply velox/benchmarks_large/TpchBenchmark.cpp]: I now combined `query` and `verbose` into a single flag. This should resolve this.

File: velox/benchmarks/tpch/TpchBenchmark.cpp [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+}
+} // namespace
+
+DEFINE_string(data_path, "", "Root path of TPC-H data");
+DEFINE_int32(
+    run_query_verbose,
+    -1,
+    "Run a given query and print execution statistics ");
+DEFINE_int32(num_drivers, 4, "Number of drivers");
+DEFINE_string(format, "parquet", "Data format");
```
Comment (mbasmanova) [inline velox/benchmarks/tpch/TpchBenchmark.cpp:60]: naming: perhaps, data_format would be clearer, e.g. match data_path better

File: velox/benchmarks/tpch/TpchBenchmark.cpp [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+            ->newConnector(kHiveConnectorId, nullptr);
+    connector::registerConnector(hiveConnector);
+  }
+
+  std::shared_ptr<Task> run(const TpchPlan& tpchPlan) {
+    CursorParameters params;
+    params.maxDrivers = FLAGS_num_drivers;
+    params.numResultDrivers = 1;
+    params.planNode = tpchPlan.plan;
+    const int kNumSplitsPerFile = FLAGS_num_splits_per_file;
```
Comment (mbasmanova) [inline velox/benchmarks/tpch/TpchBenchmark.cpp:85]: naming: kXxx names are used for static constants. Here, kNumSplitsPerFile is a regular variable, hence, the name should be numSplitsPreFile.

File: velox/benchmarks/tpch/TpchBenchmark.cpp [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+  const auto planContext = queryBuilder->getQueryPlan(18);
+  benchmark.run(planContext);
+}
+
+int main(int argc, char** argv) {
+  folly::init(&argc, &argv, false);
+  benchmark.initialize();
+  queryBuilder = std::make_shared<TpchQueryBuilder>(toFileFormat(FLAGS_format));
+  queryBuilder->initialize(FLAGS_data_path);
+  if (FLAGS_run_query_verbose == -1) {
```
Comment (mbasmanova) [inline velox/benchmarks/tpch/TpchBenchmark.cpp:131]: This flag is confusing to me. The name suggests that it allows to restrict the benchmark to just 1 query and also print out some extra information about the run. The actual meaning is different though.

**Review by mbasmanova — APPROVED** [review]
@majetideepak Overall looks good. I have some small comments below.

File: velox/benchmarks/tpch/TpchBenchmark.cpp (hunk 18 above)
Comment (majetideepak) [reply velox/benchmarks/tpch/TpchBenchmark.cpp:131]: @mbasmanova Can you please elaborate on "The actual meaning is different though."? I checked the "folly benchmark" API and there is no way to specify a particular benchmark to run. So the code here manually invokes the `benchmark.run()`.

Comment (mbasmanova) [reply velox/benchmarks/tpch/TpchBenchmark.cpp:131]: My understanding that running a benchmark via folly::runBenchmarks(); includes running the code multiple times measuring the execution time and reporting some average or min of these time. When run_query_verbose is enabled, we do not do that. Instead, we run the code just once and report the execution time from a single run. Please, correct me if my understanding is off.

Comment (majetideepak) [reply velox/benchmarks/tpch/TpchBenchmark.cpp:131]: Your understanding is correct. `run_query_verbose` only executes the query once. The folly benchmark API does not allow us to tap into the execution for a single benchmark.

Comment (majetideepak) [reply velox/benchmarks/tpch/TpchBenchmark.cpp:131]: As the `run_query_verbose` flag mentions `query` and not `benchmark` this should be okay?

File: velox/exec/tests/utils/TpchQueryBuilder.cpp [hunk 19]
```diff
@@ -280,6 +288,7 @@ TpchPlan TpchQueryBuilder::getQ18Plan() const {
               {PlanBuilder(planNodeIdGenerator)
                    .tableScan(lineitemSelectedRowType)
                    .capturePlanNodeId(lineitemScanNodeId)
+                   .project(lineitemColumnAlias)
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/exec/tests/utils/TpchQueryBuilder.cpp:291]: I tried this but it failed since the output type of the tablescan does not match the input of the project that follows.

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/tests/utils/TpchQueryBuilder.cpp:291]: You need to specify the output to match. Can you show the code that didn't work?

[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/exec/tests/utils/TpchQueryBuilder.cpp]: I made this change. Thanks for your help.

**Comment by mbasmanova** [comment]
@majetideepak Deepak, thank you for iterating on the PR. There are some macos-build failures. Would you take a look?

**Comment by mbasmanova** [comment]
@majetideepak Deepak, do you have some instructions on how to run this benchmark, specifically, what should be the layout of the data_path? 

CC: @Yuhta

**Comment by majetideepak** [comment]
@mbasmanova I wrote in the description that the layout should be hive-style partitioning. 
Example: if `--data_path=/mnt/data` then the layout will look like. It is probably best to add a comment to the code regarding this. I will add one.
```
/mnt/data/nation
/mnt/data/orders
......
```

**Comment by mbasmanova** [comment]
@majetideepak Thank you, Deepak.
